### PR TITLE
Add spaces_between_special_tokens and cleanup tokenization spaces

### PR DESCRIPTION
**Pull request by ArthurZucker**
Following [#19293](https://github.com/huggingface/transformers/issues/19293) in transformers, we are going to add the `spaces_between_special_tokens` argument to the `decode` function. 

cc @ankrgyl

### CODE REVIEW

File: bindings/python/py_src/tokenizers/implementations/base_tokenizer.py [hunk 1]
```diff
@@ -279,13 +294,20 @@ def decode_batch(self, sequences: List[List[int]], skip_special_tokens: Optional
             skip_special_tokens: (`optional`) boolean:
                 Whether to remove all the special tokens from the output strings
 
+            spaces_between_special_tokens: (`optional`) boolean:
+                Whether an extratr space (" ") should be prepended to special tokens
```
Comment (Narsil) [inline bindings/python/py_src/tokenizers/implementations/base_tokenizer.py:298]:
```suggestion
                Whether an extra space (" ") should be prepended to special tokens
```

File: tokenizers/src/tokenizer/mod.rs [hunk 2]
```diff
@@ -795,14 +795,21 @@ where
...
         let tokens = ids
             .into_iter()
             .filter_map(|id| {
                 self.added_vocabulary
                     .id_to_token(id, &self.model)
                     .filter(|token| {
-                        !skip_special_tokens || !self.added_vocabulary.is_special_token(token)
+                        !skip_special_tokens
+                            || !self.added_vocabulary.is_special_token(token)
+                            || !spaces_between_special_tokens
```
Comment (Narsil) [inline tokenizers/src/tokenizer/mod.rs:812]: Huh ?

This seems a bit odd.  spaces_between_special_tokens should control how to reconstruct the string no ?

**Review by Narsil — COMMENTED** [review]
Overall looks good.

I'm a bit surprised by the core fix.
I think adding another test with `spaces_between_special_tokens=False` should help confirm behavior.
We probably need to test against a few different decoders too. (ByteLevel and Whitespace at least).

File: tokenizers/src/tokenizer/mod.rs (hunk 2 above)
Comment (ArthurZucker) [reply tokenizers/src/tokenizer/mod.rs:812]: Yep, I think I am missing this part

**Comment by RomanCast** [comment]
Hello, any update on this PR? It would be really helpful if integrated into tokenizers! 

Thanks for the work!

**Comment by ArthurZucker** [comment]
Hey! Should be updated this week 😉

**Comment by lucasjinreal** [comment]
Please update this 1 years later.

https://github.com/huggingface/tokenizers/issues/1258

this issue indicates some REAL problems might becaused of this

File: tokenizers/src/tokenizer/mod.rs [hunk 3]
```diff
@@ -795,22 +816,59 @@ where
...
+        let sub_texts_concatenated = if spaces_between_special_tokens {
+            sub_texts.join(" ")
+        } else {
+            sub_texts.join("")
+        };
+
+        let tokens = sub_texts_concatenated
+            .split(' ')
+            .map(|s| s.to_string())
+            .collect();
```
Comment (ArthurZucker) [inline tokenizers/src/tokenizer/mod.rs:862]: Use stateless instead of this

File: tokenizers/src/tokenizer/mod.rs [hunk 4]
```diff
@@ -643,6 +643,11 @@ where
         final_vocab
     }
 
+    /// Get the added tokens
+    pub fn get_added_tokens(&self) -> Vec<String> {
+        self.added_vocabulary.get_added_tokens()
+    }
```
Comment (Narsil) [inline tokenizers/src/tokenizer/mod.rs:649]: Let's remove this. Vec allocs is costly, you don't want to do this.

File: tokenizers/src/tokenizer/added_vocabulary.rs [hunk 5]
```diff
@@ -192,6 +192,14 @@ impl AddedVocabulary {
         &self.added_tokens_map
     }
 
+    /// Get only the added tokens
+    pub fn get_added_tokens(&self) -> Vec<String> {
+        self.added_tokens
+            .iter()
+            .map(|token| token.content.clone())
+            .collect()
+    }
```
Comment (Narsil) [inline tokenizers/src/tokenizer/added_vocabulary.rs:201]: Let's remove this. This is creating an entire new vec, which isn't necessary.

File: tokenizers/tests/documentation.rs [hunk 6]
```diff
@@ -451,10 +454,31 @@ fn pipeline_bert() -> tokenizers::Result<()> {
...
+#[test]
+fn spaces_between_added_tokens() -> tokenizers::Result<()> {
+    let mut bert_tokenizer = Tokenizer::from_file("data/bert-wiki.json")?;
+    bert_tokenizer.add_special_tokens(&[
+        AddedToken::from("[ABC]", false),
+        AddedToken::from("[DEF]", false),
+        AddedToken::from("GHI IHG", false),
+    ]);
+    let input_ids = bert_tokenizer.encode("[ABC][DEF][ABC]GHI IHG[DEF]", false)?;
+    let decoded_wo_spaces = bert_tokenizer.decode(input_ids.get_ids(), true, false)?;
```
Comment (Narsil) [inline tokenizers/tests/documentation.rs:474]: Please assert the ids so we can see what we're talking about.

File: tokenizers/tests/documentation.rs [hunk 7]
```diff
@@ -451,10 +454,31 @@ fn pipeline_bert() -> tokenizers::Result<()> {
...
+    println!("{}", decoded_wo_spaces);
+    assert_eq!(decoded_wo_spaces, "[ABC][DEF][ABC]GHI IHG[DEF]");
+
+    let decoded_w_spaces = bert_tokenizer.decode(input_ids.get_ids(), true, true)?;
+    println!("{}", decoded_w_spaces);
+    // "[ABC] [DEF] [ABC] GHI IHG [DEF]"
+    assert_ne!(decoded_wo_spaces, decoded_w_spaces);
+    assert_eq!(decoded_w_spaces, "[ABC] [DEF] [ABC] GHI IHG [DEF]");
```
Comment (Narsil) [inline tokenizers/tests/documentation.rs:482]: Do you mind adding a special in the mix so we can check that skip still works with spaces between ?

File: tokenizers/src/tokenizer/mod.rs [hunk 8]
```diff
@@ -795,22 +800,37 @@ where
...
+                    continue;
+                } else if (self.added_vocabulary.get_added_tokens().contains(&token)
+                    && spaces_between_added_tokens
+                    || push_spaces)
+                    && index < last_idx - 1
+                {
+                    tokens_to_decode.extend([token, " ".to_string()]);
+                } else {
+                    tokens_to_decode.push(token);
+                }
```
Comment (Narsil) [inline tokenizers/src/tokenizer/mod.rs:827]: I'm not sure this works.

This is already tons better than before, but you're still modifying "tokens_to_decode" by introducing extra spaces.

The `push_spaces` is really convoluted (it's defined as double negative) and the you use `||` in the condition, meaning we always add if there's no decoder, but we add only if it's an added token when there's a decoder.

Refactoring the conditions names might make this a bit more obvious what we're trying to achieve.

Also adding extra "tokens" containing `" "` will *definitely* have unintended consequences in the decoders....
I think this kind of invalidates the approach.

Another idea would be to "add a decoder" at the end of the chain that would deal with those added tokens or somethign like that. In any case, "faking" tokens will yield issues I think.

File: tokenizers/tests/documentation.rs (hunk 7 above)
Comment (ArthurZucker) [reply tokenizers/tests/documentation.rs:482]: Sure!

File: tokenizers/src/tokenizer/added_vocabulary.rs (hunk 5 above)
Comment (ArthurZucker) [reply tokenizers/src/tokenizer/added_vocabulary.rs]: Removed it, totally agree

File: tokenizers/src/tokenizer/mod.rs (hunk 4 above)
Comment (ArthurZucker) [reply tokenizers/src/tokenizer/mod.rs]: yep

File: tokenizers/src/tokenizer/mod.rs (hunk 8 above)
Comment (ArthurZucker) [reply tokenizers/src/tokenizer/mod.rs]: Refactored it!

File: tokenizers/src/tokenizer/mod.rs [hunk 9]
```diff
@@ -795,22 +795,37 @@ where
...
+                if self.added_vocabulary.is_special_token(&token) && skip_special_tokens {
+                    continue;
+                }
+                if self.added_vocabulary.get_vocab().contains_key(&token) {
+                    let idx = tokens_to_decode.len();
+                    if spaces_between_added_tokens && !join_on_spaces {
+                        token += " "
+                    }
+                    if join_on_spaces && idx >= 1 && !spaces_between_added_tokens {
+                        tokens_to_decode[idx - 1] += &token;
```
[the remark this answers is not in the document]
Comment (ArthurZucker) [reply tokenizers/src/tokenizer/mod.rs:818]: The problem is that at the end, there's going to be a `tokens_to_decode.join(" ")`, thus if I don't concatenate to the previous element, spaces will be added

[the remark this answers is not in the document]
Comment (ArthurZucker) [reply tokenizers/src/tokenizer/mod.rs:818]: So I'll get rid of the join

**Comment by ArthurZucker** [comment]
The PR is ready, but after internal discussion it was decided that it will not be merged. 
The problem is that `decoders` don't match well with this, and the argument is rather not intuitive.

**Pull request closed without merging**